docs/migrations/support-slashed-denoms.md

```diff
@@ -0,0 +1,97 @@
+# Migrating from Unsupported Base Denom Slashes to Supporting Slashed BaseDenoms
+
+This document is intended to highlight significant changes which may require more information than presented in the CHANGELOG.
+Any changes that must be done by a user of ibc-go should be documented here.
+
+There are four sections based on the four potential user groups of this document:
+- Chains
+- IBC Apps
+- Relayers
+- IBC Light Clients
+
+**Note:** ibc-go supports golang semantic versioning and therefore all imports must be updated to bump the version number on major releases.
+```go
+github.com/cosmos/ibc-go/v3 -> github.com/cosmos/ibc-go/v4
+```
+
+This document is necessary when chains are upgrading from a version that does not support slashed base denoms (e.g. v3.0.0) to a version that does (e.g. v3.1.0).
+
+If a chain receives a slashed denom before it upgrades to supporting it, the receive may pass however the trace information will be incorrect.
+
+E.g. If a base denom of `testcoin/testcoin/testcoin` is sent to a chain that does not support slashes in the base denom; the receive will be successful. However, the trace information stored on the receiving chain will be: `Trace: "testcoin/testcoin", BaseDenom: "testcoin"`
+
+This incorrect trace information must be corrected when the chain does upgrade to fully supporting slashed denominations.
+
+To do so, chain binaries should include a migration script that will run when the chain upgrades from not supporting slashed base denominations to supporting slashed base denominations.
+
+## Chains
+
+### Transfer
+
+The transfer module will now support slashes in base denoms, so we must iterate over current traces to check if any of them are incorrectly formed and correct the trace information.
+
+### Upgrade Proposal
+
+```go
+// Here the upgrade name is the upgrade name set by the chain
+app.UpgradeKeeper.SetUpgradeHandler("supportSlashingDenomUpgrade",
```

Comment:
The code needed some adjustments to make it compile in the ibc-go simapp:

```golang
app.UpgradeKeeper.SetUpgradeHandler("supportSlashingDenomUpgrade",
	func(ctx sdk.Context, _ upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
		// list of traces that must replace the old traces in store
		var newTraces []ibctransfertypes.DenomTrace
		app.TransferKeeper.IterateDenomTraces(ctx,
			func(dt ibctransfertypes.DenomTrace) bool {
				// check if the new way of splitting FullDenom
				// into Trace and BaseDenom is the same as the current
				// DenomTrace.
				// If it isn't then store the new DenomTrace in the list of new traces.
				newTrace := ibctransfertypes.ParseDenomTrace(dt.GetFullDenomPath())
				if !reflect.DeepEqual(newTrace, dt) {
					newTraces = append(newTraces, newTrace)
				}
				return false
			})

		// replace the outdated traces with the new trace information
		for _, nt := range newTraces {
			app.TransferKeeper.SetDenomTrace(ctx, nt)
		}

		return app.mm.RunMigrations(ctx, app.configurator, fromVM)
	})
```

Reply:
I tested the upgrade handler locally and it works. 

Before migration:
```
denom_traces:
- base_denom: "1"
  path: transfer/channel-0/gamm/poolpoolpool
- base_denom: "1"
  path: transfer/channel-0/transfer/channel-1/gamm/testtesttest
pagination:
  next_key: null
  total: "0"
```

After migration:
```
denom_traces:
- base_denom: gamm/poolpoolpool/1
  path: transfer/channel-0
- base_denom: gamm/testtesttest/1
  path: transfer/channel-0/transfer/channel-1
pagination:
  next_key: null
  total: "0"  
```